### PR TITLE
blog: How to Migrate Existing Terraform Cloud Resources to KubeVela

### LINKED ISSUE #676: Write a blog to introduce how to immigrate your legacy Terraform cloud resources to KubeVela
Based on the implementation of https://github.com/kubevela/kubevela/issues/3672, it's better to write a blog to introduce how to immigrate your legacy Terraform cloud resources to KubeVela.

/cc @loheagn

### PR DESCRIPTION
**Pull request by loheagn**
fix #676 

Signed-off-by: loheagn <loheagn@icloud.com>

### CODE REVIEW

**Comment by loheagn** [comment]
Hi, @zzxwill , I wrote a blog to discuss how to immigrate legacy Terraform cloud resources to KubeVela :)

/cc @wonderflow

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+---
+title: How to Immigrate Legacy Terraform Cloud Resources to KubeVela
```
Comment (wonderflow) [inline blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]:
```suggestion
title: How to Migrate Existing Terraform Cloud Resources to KubeVela
```

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+image: https://raw.githubusercontent.com/oam-dev/KubeVela.io/main/docs/resources/KubeVela-03.png
+hide_table_of_contents: false
+---
+
+You may have learned from [this blog](./2022-06-27-terraform-integrate-with-vela.md) that we can use vela to manage cloud resources (like s3 bucket, AWS EIP and so on) via the terraform plugin. We can create an application which contains some cloud resource components and this application will generate these cloud resources, then we can use vela to manage them. 
+
+Sometimes we already have some Terraform cloud resources which may be created and managed by the Terraform binary or something else. We may want to import these Terraform cloud resources into KubeVela and use vela to manage them. But if we just create an application which describes these cloud resources, the cloud resources will be recreated and may lead to errors. To fix this problem, we made [a simple `backup_restore` tool](https://github.com/kubevela/terraform-controller/tree/master/hack/tool/backup_restore). This blog will show you how to use the `backup_restore` tool to import your legacy Terraform cloud resources into KubeVela.
+
```
Comment (wonderflow) [inline blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: Let's describe the benifits from doing so:

```
You can have the following benefits from doing so:

- ...
- ...
- ...
```

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md (hunk 1 above)
Comment (wonderflow) [reply blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: you should use existing "instead" of "legacy" as we regard legacy as useless thing.

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    ```
+
+3. Run the `restore` command:
+    ```shell
+    go run main.go restore --application <path/to/your/application.yaml> --component sample-s3 --state <path/to/your/state.json>
+    ```
+    The above command will resume the Terraform backend in the Kubernetes first and then create the application without recreating the S3 bucket.
+
+That's all! You have successfully immigrate the management of the S3 bucket to KubeVela!
+
```
Comment (wonderflow) [inline blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: we should describe some use cases of which we can benefit from the migration.

For example, we can update the resources, delete resources with the some lifecycle of the Application.

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md (hunk 1 above)
Comment (zzxwill) [reply blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: It's okay to say legacy like `Migrating Legacy Applications To The Cloud`.

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    ```shell
+    terraform init && terraform apply
+    ```
+
+7. Check the S3 bucket list to make sure the bucket is created successfully.
+
+## Step 2: Get the Terraform State of the Cloud Resource
+
+Run `terraform state pull` to get the Terraform state of the cloud resource and store it into a local file:
```
Comment (zzxwill) [inline blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: In Step 1, a tf.state.json file should already be generated.

**Comment by charlie0129** [comment]
We'd better use **migrate** instead of **immigrate**.

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+Sometimes we already have some Terraform cloud resources which may be created and managed by the Terraform binary or something else. We may want to import these Terraform cloud resources into KubeVela and use vela to manage them. But if we just create an application which describes these cloud resources, the cloud resources will be recreated and may lead to errors. To fix this problem, we made [a simple `backup_restore` tool](https://github.com/kubevela/terraform-controller/tree/master/hack/tool/backup_restore). This blog will show you how to use the `backup_restore` tool to import your legacy Terraform cloud resources into KubeVela.
+
+## Step 1: Create Legacy Terraform Cloud Resources
+
+Since we are going to demonstrate how to import an existing cloud resource into KubeVela, we need to create one first. If you already have such resources, you can skip this step.
+
+Before start, make sure you have:
+
+- Installed [terraform CLI](https://www.terraform.io/downloads).
+- Have a Cloud Service credentials, in this article, we will use Alibaba Cloud as example.
```
Comment (chivalryq) [inline blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]:
```suggestion
- Have a Cloud Service credentials, in this article, we will use aws as example.
```

**Comment by loheagn** [comment]
> we should describe some use cases of which we can benefit from the migration.
> For example, we can update the resources, delete resources with the some lifecycle of the Application.

Hi, @wonderflow sorry for the late reply. I found that "the benefits of the migration" are roughly the same as "the benefits of using vela to manage the terraform cloud resources". So I use a link to show the benefits. You can review it [here](https://github.com/kubevela/kubevela.io/pull/833/files#diff-5c19091a47aabee70843d202a5958f0d42e9fc115760a0649ee4d8609dde33b1R14) :)

File: blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md (hunk 5 above)
Comment (loheagn) [reply blog/2022-07-04-immigrate-erraform-cloud-resources-to-kubeVela.md]: Thanks. I have fixed it.

File: blog/2022-07-20-migrate-erraform-cloud-resources-to-kubeVela.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    spec:
+    components:
+        - name: sample-s3
+        type: aws-s3
+        properties:
+            bucket: vela-website-202110191745
+            acl: private
+
+            writeConnectionSecretToRef:
+            name: s3-conn
```
Comment (chivalryq) [inline blog/2022-07-20-migrate-erraform-cloud-resources-to-kubeVela.md]:
```suggestion
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: app-aws-s3
spec:
  components:
    - name: sample-s3
      type: aws-s3
      properties:
        bucket: vela-website-202110191745
        acl: private
        writeConnectionSecretToRef:
        name: s3-conn
```

**Review by chivalryq — CHANGES_REQUESTED** [review]
Note the format

**Comment by chivalryq** [comment]
Generally LGTM, fix the format and let's merge this

**Review by chivalryq — APPROVED** [review]

**Pull request merged**